pydantic/_internal/_decorators.py

```diff
@@ -774,7 +778,25 @@ def get_function_return_type(
 
 
 def count_positional_params(sig: Signature) -> int:
-    return sum(1 for param in sig.parameters.values() if can_be_positional(param))
+    """Get the number of positional (required) arguments of a signature.
+
+    This function should only be used to inspect signatures of validation and serialization functions.
```

Comment:
I'm not too happy with how this function grew, hence this "warning" in the docstring. At some point, maybe this inspection of validators/serializers signature could be refactored, e.g.:

```python
@dataclass
class FunctionInfo:
    func: Callable[..., Any]

    @property
    def requires_info_arg(self) -> bool: ...

    ...
```